Register block style early.

spacedmonkey opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/58528
---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

spacedmonkey committed:
Register block style early.

spacedmonkey committed:
Move logic

spacedmonkey committed:
remove logic.

spacedmonkey committed:
Cleaner logic.

westonruter commented on src/wp-includes/blocks/index.php:
```diff
@@ -19,6 +19,46 @@
+			if ( is_array( $style_handle ) ) {
+				continue;
+			}
+			$path = "/wp-includes/blocks/$name/style$suffix.css";
```
```suggestion
			$path = "/wp-includes/blocks/$name/style{$suffix}.css";
```

westonruter commented on src/wp-includes/blocks/index.php:
```diff
@@ -19,6 +19,46 @@
+	foreach ( $core_blocks_meta as $name => $schema ) {
+		foreach ( $style_fields as $style_field ) {
+			if ( ! isset( $schema[ $style_field ] ) ) {
+				$style_handle = 'style' === $style_field ? "wp-block-$name" : "wp-block-{$name}-editor";
```
```suggestion
				$style_handle = 'style' === $style_field ? "wp-block-{$name}" : "wp-block-{$name}-editor";
```

westonruter reviewed (commented):
Just a couple driveby nits

spacedmonkey committed:
Apply suggestions from code review

Co-authored-by: Weston Ruter <westonruter@google.com>

spacedmonkey committed:
Break into it's own action.

spacedmonkey commented:
CC @aristath

spacedmonkey committed:
Merge branch 'trunk' into fix/realpath-error-5

felixarntz commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,54 @@
+
+	foreach ( $core_blocks_meta as $name => $schema ) {
+		foreach ( $style_fields as $style_field ) {
+			if ( ! isset( $schema[ $style_field ] ) ) {
+				$style_handle = 'style' === $style_field ? "wp-block-{$name}" : "wp-block-{$name}-editor";
+				wp_register_style(
+					$style_handle,
+					false
+				);
+				continue;
+			}
```
Why is this needed? In `register_block_style_handle()`, if the field doesn't exist in the block metadata, it just returns `false` and doesn't call `wp_register_style()`.

The logic used to potentially register with `false` URI is different in `register_block_style_handle()`, we should make sure to follow that here.

spacedmonkey replied:
This confused me as well, but consider. https://github.com/WordPress/wordpress-develop/commit/716a99587c04e5b37b4ceacc3ccce6c72c842464 / [https://core.trac.wordpress.org/ticket/53358 #53358] / https://github.com/WordPress/gutenberg/pull/32275/. 

Basically what happens in `register_block_style_handle` use realpath to check if the file exists. Look at this snippet 

https://github.com/WordPress/wordpress-develop/blob/fbef5030f67b6837119e9c890288bb0ace796940/src/wp-includes/blocks.php#L225-L255

If the file doesn't exist, then sets `$style_uri` to false. This is so that developers can add inline styles to these style sheets and style core blocks this way. Core itself also use these styles, see wp_common_block_scripts_and_styles. Example 

https://github.com/WordPress/wordpress-develop/blob/fbef5030f67b6837119e9c890288bb0ace796940/src/wp-includes/script-loader.php#L2371

With this, lots of themes / plugins and core itself would break. It is confusing, but it is needed. Maybe we should add a comment explaining it...

felixarntz replied:
@spacedmonkey I understand, however I am referring more specifically to the exact logic: There are three possible scenarios technically:
1. The `style`/`editorStyle` properties are defined in the schema _and_ the files exist.
2. The `style`/`editorStyle` properties are defined in the schema, but the files do not exist.
3. The `style`/`editorStyle` properties are not defined in the schema.

Right now, the PR registers the stylesheet with `false` in case 3 above. However, `register_block_style_handle()` does nothing in case 3, it bails early. Instead, it registers the stylesheet with `false` in case 2 above. So that is where your implementation and the current one in core differ.

I am not entirely sure how relevant that is for core blocks. Can you do some testing with your PR and check which core blocks (if any) the condition in https://github.com/WordPress/wordpress-develop/pull/4621/files#diff-c62beff71446448ff952e38805c8ae7652f4e2717d3ca41b1d9cd9963bd82464R44 applies to? That would help clarify whether this code is okay as is or whether it's safer to remove it.

felixarntz commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,54 @@
+				continue;
+			}
+			$path = "/wp-includes/blocks/$name/style{$suffix}.css";
+			wp_register_style( $style_handle, $path );
```
This needs to use the URL like in `register_block_style_handle()` instead of path:
```suggestion
			wp_register_style( $style_handle, includes_url( "blocks/$name/style{$suffix}.css" );
```

spacedmonkey replied:
If you look at `wp_default_scripts`, other core styles do not do this. I am not sure why, but they do not. It is worth noting that most of these styles will never use this url for anything, as they will be inlines by `wp_maybe_inline_styles`. 

I do see why this should different from `wp_default_scripts`, but I could be talked out of it.

felixarntz replied:
Better to keep this like it was before / like it is in `register_block_style_handle()`. I see you updated it, so all good now 👍

felixarntz commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,54 @@
+ *
+ * @since 6.3.0
+ */
+function register_core_block_types_styles() {
```
I think a better name would be `register_core_block_style_handles()`, because this function basically does what otherwise `register_block_style_handle()` would do, only in a more efficient way for core blocks.

felixarntz commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,54 @@
 require BLOCKS_PATH . 'require-dynamic-blocks.php';
 
+/**
+ * Registers core block types styles.
```
It's worth clarifying why this function exists, compared to `register_block_style_handle()`.
```suggestion
 * Registers core block style handles.
 *
 * While {@see register_block_style_handle()} is typically used for that, the way it is
 * implemented is inefficient for core block styles. Registering those style handles here
 * avoids unnecessary logic and filesystem lookups in the other function.
```

felixarntz reviewed (changes requested):
@spacedmonkey This looks great so far. Only a bit of feedback below.

spacedmonkey committed:
Feedback.

spacedmonkey committed:
Merge branch 'trunk' into fix/realpath-error-5

spacedmonkey committed:
More feedback

mukeshpanchal27 commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,62 @@
+			if ( is_array( $style_handle ) ) {
+				continue;
+			}
+			$path = "/wp-includes/blocks/$name/style{$suffix}.css";
```
```suggestion
			$path = "/wp-includes/blocks/{$name}/style{$suffix}.css";
```

mukeshpanchal27 commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,62 @@
+		$core_blocks_meta = require ABSPATH . WPINC . '/blocks/blocks-json.php';
+	}
+
+	$suffix = SCRIPT_DEBUG ? '' : '.min';
```
```suggestion
	$suffix = wp_scripts_get_suffix();
```
Can we use function instead of constant as core use function in many files.

mukeshpanchal27 commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,62 @@
+				$wp_styles->add_data( $style_handle, 'suffix', $suffix );
+				$wp_styles->add_data( $style_handle, 'path', $rtl_file );
+			}
+
```
```suggestion
```

mukeshpanchal27 reviewed (commented):
Great work @spacedmonkey, Left some feedback.

spacedmonkey committed:
Apply suggestions from code review

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

mukeshpanchal27 commented on src/wp-includes/blocks.php:
```diff
@@ -205,6 +205,10 @@ function register_block_style_handle( $metadata, $field_name, $index = 0 ) {
 		$style_handle = $style_handle[ $index ];
 	}
 
+	if ( wp_style_is( $style_handle, 'registered' ) ) {
```
@spacedmonkey, could you please add a small description of this change so that others can understand it as well?

mukeshpanchal27 commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,61 @@
+			if ( is_array( $style_handle ) ) {
+				continue;
+			}
+			$path = "/wp-includes/blocks/{$name}/style{$suffix}.css";
+			$wp_styles->add( $style_handle, $path );
+			$wp_styles->add_data( $style_handle, 'path', ABSPATH . $path );
```
```suggestion
			$path = WPINC . "/blocks/{$name}/style{$suffix}.css";
			$wp_styles->add( $style_handle, "/{$path}" );
			$wp_styles->add_data( $style_handle, 'path', ABSPATH . $path );
```

`ABSPATH . $path` added duplicate forward slash. Let's follow similar approach that used here https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/script-loader.php#L1600-L1602

**Image:**
<img width="938" alt="Screenshot 2023-06-20 at 10 18 50 AM" src="https://github.com/WordPress/wordpress-develop/assets/10103365/42b8ce7f-8c65-4c5d-acfc-4915a3fa6cb3">

mukeshpanchal27 commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,61 @@
+			$wp_styles->add( $style_handle, $path );
+			$wp_styles->add_data( $style_handle, 'path', ABSPATH . $path );
+
+			$rtl_file = str_replace( "{$suffix}.css", "-rtl{$suffix}.css", ABSPATH . $path );
```
Could we move `str_replace` code inside the `is_rtl()` condition?  this way we reduce 184 replacement of style for normal sites and only perform it for RTL sites.

spacedmonkey replied:
This code is copied from elsewhere. I think we should leave as is. 

https://github.com/WordPress/wordpress-develop/blob/76376fdbc3dc0b3261de377dffc350677345e7ba/src/wp-includes/blocks.php#L263-L273

mukeshpanchal27 reviewed (commented):
@spacedmonkey A few last feedback, but almost good to go.

spacedmonkey committed:
Merge branch 'trunk' into fix/realpath-error-5

spacedmonkey committed:
Feedback.

spacedmonkey committed:
Reusable.

mukeshpanchal27 commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,63 @@
+			}
+			$style_path = "blocks/{$name}/style{$suffix}.css";
+			$wp_styles->add( $style_handle, $includes_url . $style_path );
+			$path = ABSPATH . WPINC . DIRECTORY_SEPARATOR . $style_path;
```
@spacedmonkey Core avoid to use `DIRECTORY_SEPARATOR` constant. See https://core.trac.wordpress.org/ticket/44654

spacedmonkey replied:
```suggestion
			$path = ABSPATH . WPINC . '/' . $style_path;
```

mukeshpanchal27 commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,63 @@
+		$core_blocks_meta = require ABSPATH . WPINC . '/blocks/blocks-json.php';
+	}
+
+	$includes_url = includes_url();
```
```suggestion
	$includes_url = includes_url( '', 'relative' );
```
Not blocker. before this changes it was `/wp-includes/blocks/blockbame/style.css, if we use `includes_url()` it return include URL link.

spacedmonkey replied:
This was added becuase of feedback from @felixarntz https://github.com/WordPress/wordpress-develop/pull/4621#discussion_r1232574807

I don't see how it hurts to include the url here.

mukeshpanchal27 reviewed (commented):
@spacedmonkey Left some feedback.

spacedmonkey committed:
Apply suggestions from code review

mukeshpanchal27 reviewed (approved):
Thanks @spacedmonkey for the update. It look good to me.

mukeshpanchal27 commented:
@spacedmonkey Re-run the failed job it show `Navigation timeout of 30000 ms exceeded`.

spacedmonkey committed:
Merge branch 'trunk' into fix/realpath-error-5

felixarntz reviewed (commented):
@spacedmonkey Looks good except for the pending conversation where I just left a follow-up on https://github.com/WordPress/wordpress-develop/pull/4621#discussion_r1237444923. Maybe it's okay as is, but we have to validate that, since right now your logic is a bit different from the existing one. So we have to make sure we don't accidentally introduce a regression there.

spacedmonkey committed:
Merge branch 'trunk' into fix/realpath-error-5

spacedmonkey committed:
Try something.

spacedmonkey committed:
Use a transinet as a cache

spacedmonkey commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,77 @@
+		'editorStyle',
+	);
+	$transient_name = 'block_styles';
+	$files          = get_transient( $transient_name );
+	if ( ! $files ) {
+		$files = glob( __DIR__ . "/**/style*{$suffix}.css" );
+		set_transient( $transient_name, $files );
+	}
```
This checks to see if the file exists once and then caches.

spacedmonkey committed:
Merge branch 'trunk' into fix/realpath-error-5

felixarntz commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,77 @@
+			}
+
+			$style_path = "blocks/{$name}/style{$suffix}.css";
+			$path       = ABSPATH . WPINC . '/' . $style_path;
```
Nit-pick: Maybe defined a `$includes_path = ABSPATH . WPINC . '/'` variable outside of the loop to improve readability. Not critical, but you already have `$includes_url`, so this would be nice for parity.

felixarntz commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,77 @@
+		'style',
+		'editorStyle',
+	);
+	$transient_name = 'block_styles';
```
Since the styles are all from core itself, I think we can easily handle invalidation here when WordPress is updated: Either we add `$wp_version` to the transient name, or we delete the transient on every WordPress update (like you're doing above). In order to support programmatically controlled sites (e.g. that update via Composer), it may be more reliable to rely on the `$wp_version` global.

Also, let's prefix this with something like `wp_core_`, to clarify that it is only for WP core block styles.
```suggestion
	$transient_name = 'wp_core_block_styles_' . $GLOBALS['wp_version'];
```

spacedmonkey replied:
@felixarntz I originally had it like this. But as the tranient would be stored in options. This would mean that a version of the transient would be stored in an autoloaded option forever more right?

felixarntz replied:
@spacedmonkey Fair point, in that case, let's just change to 
```suggestion
	$transient_name = 'wp_core_block_styles';
```
and use invalidation on update.

felixarntz commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,77 @@
+	}
+
+	foreach ( $core_blocks_meta as $name => $schema ) {
+		if ( ! isset( $schema['style'] ) ) {
+			$schema['style'] = "wp-block-$name";
+		}
+		if ( ! isset( $schema['editorStyle'] ) ) {
+			$schema['editorStyle'] = "wp-block-{$name}-editor";
+		}
```
Before this backfilling, we also need to run the `block_type_metadata` filter, otherwise the data may become inconsistent.
```suggestion
		/** This filter is documented in wp-includes/blocks.php */
		$schema = apply_filters( 'block_type_metadata', $schema );

		// Backfill these properties similar to `register_block_type_from_metadata()`.
		if ( ! isset( $schema['style'] ) ) {
			$schema['style'] = "wp-block-$name";
		}
		if ( ! isset( $schema['editorStyle'] ) ) {
			$schema['editorStyle'] = "wp-block-{$name}-editor";
		}
```

felixarntz reviewed (commented):
@spacedmonkey This is pretty much good to go, after another review the changes make sense for parity with the existing `register_block_style_handle()`. A few outstanding points of feedback below.

Note that I found a bug that is related to this code, see https://core.trac.wordpress.org/ticket/58605.

spacedmonkey committed:
Update src/wp-includes/blocks/index.php

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

spacedmonkey committed:
More glob.

spacedmonkey committed:
Feedback.

spacedmonkey committed:
Merge branch 'trunk' into fix/realpath-error-5

spacedmonkey commented:
@felixarntz @mukeshpanchal27 I went down a habbit hole. I fixed two more issues. First there is a fix for [https://core.trac.wordpress.org/ticket/58605](https://core.trac.wordpress.org/ticket/58605). Changing the glob, fixes this. 
Secondly, looking at the function `wp_common_block_scripts_and_styles`. This too does a glob on css files as well. It doesn't make sense to do a glob in two places. So I moved this logic to this function as well. This has some really nice saves as well. To test this, you should switch to the tt2 theme. 

This PR now needs some unit tests, I will work on the ASAP.

spacedmonkey committed:
Only add new style if supported

felixarntz commented:
@spacedmonkey I don't think it's a good idea to intertwine the two tickets in this pull request. Yes, they are related but we're also close to the 6.3 beta, and you now added lots of new logic that changed the PR notably from what it was before.

I would have approved it now in its original scope, but the new changes make this a whole lot more complicated and potentially risky since you moved logic from one place to the other. I would suggest that you revert those changes and open a separate PR to address https://core.trac.wordpress.org/ticket/58560. That ticket is much less critical, and it's not even in the 6.3 milestone. Let's tackle it separately. Then we can get this PR merged in its original scope.

spacedmonkey committed:
Revert wp_common_block_script_and_styles changes.

spacedmonkey commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,82 @@
+		'editorStyle' => 'editor',
+	);
+	$transient_name = 'wp_core_block_styles';
+	$files          = get_transient( $transient_name );
+	if ( ! $files ) {
+		$files = glob( __DIR__ . "/**/**{$suffix}.css" );
+		set_transient( $transient_name, $files );
```
@felixarntz Should we disable this cache is the WP_DEBUG is enabled?

felixarntz replied:
@spacedmonkey Sounds good. Please add the same `todo` comment that is present on all the other places where we do caching around e.g. the `theme.json` stuff.

Related: https://core.trac.wordpress.org/ticket/57487

spacedmonkey replied:
@felixarntz Pushed the change.

spacedmonkey commented:
@felixarntz Fair point, adding back is simple enough. I will do it as part of that release. I have reverted the changes now, can you take a look again.

felixarntz commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,82 @@
+			$style_path = "blocks/{$name}/{$filename}{$suffix}.css";
+			$path       = $includes_path . $style_path;
+
+			if ( ! in_array( $path, $files, true ) ) {
+				$wp_styles->add(
+					$style_handle,
+					false
+				);
+			} else {
+				$wp_styles->add( $style_handle, $includes_url . $style_path );
+				$wp_styles->add_data( $style_handle, 'path', $path );
+
+				$rtl_file = str_replace( "{$suffix}.css", "-rtl{$suffix}.css", $path );
+				if ( is_rtl() && in_array( $rtl_file, $files, true ) ) {
+					$wp_styles->add_data( $style_handle, 'rtl', 'replace' );
+					$wp_styles->add_data( $style_handle, 'suffix', $suffix );
+					$wp_styles->add_data( $style_handle, 'path', $rtl_file );
+				}
+			}
```
Using if-else is not a great pattern, and we could easily avoid it here. I think you previously had it like that, not sure why it changed. Better for readability to use early returns than nesting.
```suggestion
			if ( ! in_array( $path, $files, true ) ) {
				$wp_styles->add(
					$style_handle,
					false
				);
				continue;
			}

			$wp_styles->add( $style_handle, $includes_url . $style_path );
			$wp_styles->add_data( $style_handle, 'path', $path );

			$rtl_file = str_replace( "{$suffix}.css", "-rtl{$suffix}.css", $path );
			if ( is_rtl() && in_array( $rtl_file, $files, true ) ) {
				$wp_styles->add_data( $style_handle, 'rtl', 'replace' );
				$wp_styles->add_data( $style_handle, 'suffix', $suffix );
				$wp_styles->add_data( $style_handle, 'path', $rtl_file );
			}
```

spacedmonkey replied:
Fixed.

felixarntz reviewed (commented):
@spacedmonkey Thanks for reverting, this is pretty much good to go. Only outstanding bit is https://github.com/WordPress/wordpress-develop/pull/4621/files#r1240041002, plus I have one nit-pick below - not a blocker, but it would be good to change.

spacedmonkey committed:
Disable cache on `WP_DEBUG`.

spacedmonkey committed:
Feedback.

spacedmonkey committed:
Do not add $suffix to glob, so it gets all css files. Even when `SCRIPT_DEBUG` is setup.

felixarntz commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,93 @@
+		'style'       => 'style',
+		'editorStyle' => 'editor',
+	);
+	$files         = array();
+
+	/*
+	 * Ignore transient cache when `WP_DEBUG` is enabled. Why? To avoid interfering with
+	 * the core developer's workflow.
+	 *
+	 * @todo Replace `WP_DEBUG` once an "in development mode" check is available in Core.
+	 */
+	if ( ! WP_DEBUG ) {
+		$transient_name = 'wp_core_block_styles';
+		$files          = get_transient( $transient_name );
+		if ( ! $files ) {
+			$files = glob( __DIR__ . "/**/**.css" );
+			set_transient( $transient_name, $files );
+		}
+	}
```
I think you have a bug here now, if `WP_DEBUG` is enabled, `$files` will always be an empty array.
```suggestion

	/*
	 * Ignore transient cache when `WP_DEBUG` is enabled. Why? To avoid interfering with
	 * the core developer's workflow.
	 *
	 * @todo Replace `WP_DEBUG` once an "in development mode" check is available in Core.
	 */
	if ( ! WP_DEBUG ) {
		$transient_name = 'wp_core_block_styles';
		$files          = get_transient( $transient_name );
		if ( ! $files ) {
			$files = glob( __DIR__ . "/**/**.css" );
			set_transient( $transient_name, $files );
		}
	} else {
		$files = glob( __DIR__ . "/**/**.css" );
	}
```

felixarntz reviewed (commented):
@spacedmonkey One last thing, I think there's a bug in the new code you added.

felixarntz commented:
@spacedmonkey FYI I just committed https://core.trac.wordpress.org/changeset/56005, which you have already accounted for here in https://github.com/WordPress/wordpress-develop/pull/4621/files#diff-c62beff71446448ff952e38805c8ae7652f4e2717d3ca41b1d9cd9963bd82464R38-R41 👍 

Maybe refresh your PR branch against `trunk` to make sure tests still pass.

spacedmonkey committed:
Update src/wp-includes/blocks/index.php

Co-authored-by: Felix Arntz <felixarntz@users.noreply.github.com>

spacedmonkey committed:
Merge branch 'trunk' into fix/realpath-error-5

spacedmonkey committed:
Revert committed file.

spacedmonkey commented on src/wp-includes/blocks.php:
```diff
@@ -368,6 +373,7 @@ function register_block_type_from_metadata( $file_or_folder, $args = array() ) {
 		if ( ! isset( $metadata['style'] ) ) {
 			$metadata['style'] = "wp-block-$block_name";
 		}
+
```
```suggestion
```

spacedmonkey replied:
```suggestion
```

felixarntz reviewed (approved):
Thanks @spacedmonkey, LGTM in terms of production code!

If you can add a few tests, that would be excellent 🎉

mukeshpanchal27 commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,94 @@
+
+		// Backfill these properties similar to `register_block_type_from_metadata()`.
+		if ( ! isset( $schema['style'] ) ) {
+			$schema['style'] = "wp-block-$name";
```
```suggestion
			$schema['style'] = "wp-block-{$name}";
```

spacedmonkey replied:
```suggestion
			$schema['style'] = "wp-block-{$name}";
```

mukeshpanchal27 reviewed (approved):
Thanks @spacedmonkey, Great work. Approved with one nit-pick.

spacedmonkey committed:
Merge branch 'trunk' into fix/realpath-error-5

spacedmonkey committed:
Improve performance and add unit tests.

spacedmonkey committed:
Merge branch 'trunk' into fix/realpath-error-5

spacedmonkey committed:
Apply suggestions from code review

spacedmonkey commented:
@felixarntz @mukeshpanchal27 As requested, I have added unit tests. The unit tests have highlighted that the incorrectly value was being return. I moved the some the logic in `register_block_style_handle` around. This does two things, one skips the need to call `wp_should_load_separate_core_block_assets`, which resulted in a 1ms of load time. Two, calls `generate_block_asset_handle` and gets the correct style handle name early, so it can be used with `wp_style_is`. 

This change now is even better of a performance improvement.

spacedmonkey committed:
Merge branch 'trunk' into fix/realpath-error-5

spacedmonkey committed:
Fix lints.

spacedmonkey committed:
Fix comment

felixarntz commented on tests/phpunit/tests/blocks/registerCoreBlockStyleHandles.php:
```diff
@@ -0,0 +1,122 @@
+
+		$core_blocks_meta = $this->get_block_data();
+
+		foreach ( $core_blocks_meta as $name => $schema ) {
```
Instead of a `foreach`, this test should use a data provider. You can change the `get_block_data()` method to become such a data provider, e.g. `data_get_core_blocks()`, and then it returns an array, which for every block includes the `$name` and `$schema`.

felixarntz commented on tests/phpunit/tests/blocks/registerCoreBlockStyleHandles.php:
```diff
@@ -0,0 +1,122 @@
+		$wp_styles        = $GLOBALS['wp_styles'];
+		$includes_url     = includes_url();
+
+		foreach ( $core_blocks_meta as $name => $schema ) {
```
Same as above, this should use a data provider method instead of having lots of assertions in a single test run going through a loop. You can use the same data provider method for this and the above test method.

felixarntz commented on tests/phpunit/tests/blocks/registerCoreBlockStyleHandles.php:
```diff
@@ -0,0 +1,122 @@
+
+		foreach ( $core_blocks_meta as $name => $schema ) {
+
+			if ( ! isset( $schema['style'] ) ) {
+				$schema['style'] = "wp-block-$name";
+			}
+			if ( ! isset( $schema['editorStyle'] ) ) {
+				$schema['editorStyle'] = "wp-block-{$name}-editor";
+			}
```
Another benefit of using a data provider is that you'll be able to do this in the data provider method rather than duplicated across both tests.

felixarntz reviewed (commented):
@spacedmonkey Production code LGTM! For the tests, please use a data provider rather than a big `foreach` loop in the tests themselves.

spacedmonkey committed:
Convert code to use a dataProvider.

westonruter commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,94 @@
+	 * @todo Replace `WP_DEBUG` once an "in development mode" check is available in Core.
+	 */
+	if ( ! WP_DEBUG ) {
+		$transient_name = 'wp_core_block_styles';
```
Would this be better named:

```suggestion
		$transient_name = 'wp_core_block_css_files';
```

westonruter commented on src/wp-admin/includes/upgrade.php:
```diff
@@ -654,6 +654,8 @@ function wp_upgrade() {
 			update_site_meta( get_current_blog_id(), 'db_last_updated', microtime() );
 		}
 
+		delete_transient( 'wp_core_block_styles' );
```
```suggestion
		delete_transient( 'wp_core_block_css_files' );
```

westonruter replied:
Also, per below, wouldn't site transients be more appropriate?

```suggestion
		delete_site_transient( 'wp_core_block_css_files' );
```

spacedmonkey replied:
@westonruter Using network options ( tranisents ) has a performance impact. I think network options are not autoloaded. So it result in a database query, slow things down. Maybe we could work this out later.

westonruter commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,94 @@
+		$files          = get_transient( $transient_name );
+		if ( ! $files ) {
+			$files = glob( __DIR__ . '/**/**.css' );
+			set_transient( $transient_name, $files );
```
Seems like this could be setting a site transient instead:

```suggestion
			set_site_transient( $transient_name, $files );
```

spacedmonkey replied:
See https://github.com/WordPress/wordpress-develop/pull/4621/files#r1242531400

spacedmonkey committed:
Workaround.

spacedmonkey committed:
Change transient name.

felixarntz reviewed (approved):
Thanks @spacedmonkey, looks great! Good to commit IMO.

felixarntz commented on src/wp-includes/blocks/index.php:
```diff
@@ -12,6 +12,94 @@
+		'editorStyle' => 'editor',
+	);
+
+	/*
+	 * Ignore transient cache when `WP_DEBUG` is enabled. Why? To avoid interfering with
+	 * the core developer's workflow.
+	 *
+	 * @todo Replace `WP_DEBUG` once an "in development mode" check is available in Core.
+	 */
+	if ( ! WP_DEBUG ) {
```
Now that https://github.com/WordPress/wordpress-develop/commit/4a16702090984caef24af5a07d598cc5afff2fdc has been committed, can you please refresh this PR against latest `trunk` and replace this?
```suggestion
	/*
	 * Ignore transient cache when the development mode is set to 'core', so it doesn't interfere with the core
	 * developer's workflow.
	 */
	if ( wp_get_development_mode() !== 'core' ) {
```

felixarntz reviewed (commented):
@spacedmonkey One small tweak here is needed now that `wp_get_development_mode()` is available in `trunk`, see below. Can you please update the code as suggested to get rid of the todo comment?

It would be nice to add a simple test to ensure that the transient is used only when the development mode is not "core", so if you can implement that before the beta release, that would be awesome, though not a blocker IMO. For reference, it could be similar to e.g. https://github.com/WordPress/wordpress-develop/commit/4a16702090984caef24af5a07d598cc5afff2fdc#diff-a2793e725e03902d8a665c105916351677858253a237e4d08bdc1c01f8c3d671R38-R61.

spacedmonkey committed:
Use `wp_get_development_mode`.

spacedmonkey committed:
Merge branch 'trunk' into fix/realpath-error-5

spacedmonkey commented:
Committed https://github.com/WordPress/wordpress-develop/commit/d8409a205a40a9c5a69d668a9aa2aa3123f4c3e5
